cmd/cert-checker/main.go

```diff
@@ -0,0 +1,261 @@
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/cactus/go-statsd-client/statsd"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/jmhodges/clock"
+	gorp "github.com/letsencrypt/boulder/Godeps/_workspace/src/gopkg.in/gorp.v1"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/policy"
+	"github.com/letsencrypt/boulder/sa"
+)
+
+const (
+	good = "valid"
+	bad  = "invalid"
+
+	filenameLayout = "20060102"
+
+	checkPeriod = time.Hour * 24 * 90
+
+	batchSize = 1000
+)
+
+type report struct {
+	begin     time.Time
+	end       time.Time
+	GoodCerts int64
+	BadCerts  int64
+	Entries   map[string]reportEntry
+}
+
+func (r *report) save(directory string) error {
+	filename := path.Join(directory, fmt.Sprintf(
+		"%s-%s-report.json",
+		r.begin.Format(filenameLayout),
+		r.end.Format(filenameLayout),
+	))
+	content, err := json.Marshal(r)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, content, os.ModePerm)
+}
+
+type reportEntry struct {
+	Valid    bool     `json:"validity"`
+	Problems []string `json:"problem,omitempty"`
+}
+
+type certChecker struct {
+	pa           core.PolicyAuthority
+	dbMap        *gorp.DbMap
+	certs        chan core.Certificate
+	issuedReport report
+	clock        clock.Clock
+}
+
+func newChecker(dbMap *gorp.DbMap) certChecker {
+	c := certChecker{
+		pa:    policy.NewPolicyAuthorityImpl(),
+		dbMap: dbMap,
+		certs: make(chan core.Certificate, batchSize),
+		clock: clock.Default(),
+	}
+	c.issuedReport.Entries = make(map[string]reportEntry)
+	return c
+}
+
+func (c *certChecker) getCerts() error {
+	c.issuedReport.begin = c.clock.Now()
+	c.issuedReport.end = c.issuedReport.begin.Add(-checkPeriod)
+
+	var count int
+	err := c.dbMap.SelectOne(
+		&count,
+		"SELECT count(*) FROM certificates WHERE issued >= :issued",
+		map[string]interface{}{"issued": c.issuedReport.end},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve certs in batches of 1000 (the size of the certificate channel)
+	// so that we don't eat unnecessary amounts of memory and avoid the 16MB MySQL
+	// packet limit.
+	// TODO(#701): This query needs to make better use of indexes
+	for offset := 0; offset < count; {
+		var certs []core.Certificate
+		_, err = c.dbMap.Select(
+			&certs,
+			"SELECT * FROM certificates WHERE issued >= :issued ORDER BY issued ASC LIMIT :limit OFFSET :offset",
+			map[string]interface{}{"issued": c.issuedReport.end, "limit": batchSize, "offset": offset},
+		)
+		if err != nil {
+			return err
+		}
+		for _, cert := range certs {
+			c.certs <- cert
+		}
+		offset += len(certs)
+	}
+
+	// Close channel so range operations won't block once the channel empties out
+	close(c.certs)
```

Comment:
So the problem with this close is that you'll race the close of the channel with items getting pulled off the channel by the workers. That is, closing a channel with items still in it will lose all of those items.

See https://play.golang.org/p/34fElBrUBn